h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/matrix/SparseMatrixFactory.java

```diff
@@ -161,30 +154,37 @@ protected void map(int chunkIdx) {
             NestedArrayPointer rowHeaderPointer = new NestedArrayPointer(rwRow);
             NestedArrayPointer dataPointer = new NestedArrayPointer(nonZeroCount);
 
-            for(long i = _f.anyVec().espc()[chunk]; i < _f.anyVec().espc()[chunk+1]; i++) {
-                if (_w != null && _w.at(i) == 0) continue;
+            Chunk weightChunk = _w != null ? _w.chunkForChunkIdx(chunk) : null;
+            Chunk respChunk = _respVec.chunkForChunkIdx(chunk);
+            Chunk[] featChunks = new Chunk[_vecs.length];
+            for (int i = 0; i < featChunks.length; i++) {
+                featChunks[i] = _vecs[i].chunkForChunkIdx(chunk);
+            }
+            
+            for(int i = 0; i < respChunk._len; i++) {
+                if (weightChunk != null && weightChunk.atd(i) == 0) continue;
                 rowHeaderPointer.setAndIncrement(_matrix._rowHeaders, nonZeroCount);
                 _actualRows[chunkIdx]++;
                 for (int j = 0; j < _di._cats; ++j) {
                     dataPointer.set(_matrix._sparseData, 1);
-                    if (_vecs[j].isNA(i)) {
+                    if (featChunks[j].isNA(i)) {
                         dataPointer.set(_matrix._colIndices, _di.getCategoricalId(j, Float.NaN));
                     } else {
-                        dataPointer.set(_matrix._colIndices, _di.getCategoricalId(j, _vecs[j].at8(i)));
+                        dataPointer.set(_matrix._colIndices, _di.getCategoricalId(j, featChunks[j].at8(i)));
                     }
                     dataPointer.increment();
                     nonZeroCount++;
                 }
                 for (int j = 0; j < _di._nums; ++j) {
-                    float val = (float) _vecs[_di._cats + j].at(i);
+                    float val = (float) featChunks[_di._cats + j].atd(i);
                     if (val != 0) {
                         dataPointer.set(_matrix._sparseData, val);
                         dataPointer.set(_matrix._colIndices, _di._catOffsets[_di._catOffsets.length - 1] + j);
                         dataPointer.increment();
                         nonZeroCount++;
                     }
                 }
-                rwRow = setResponseAndWeight(_w, _resp, _weights, _respVec, rwRow, i);
+                rwRow = setResponseAndWeight(_w != null ? _w.new Reader() : null, _resp, _weights, _respVec.new Reader(), rwRow, i);
```

Comment:
there is an alternative method that deals with vecs directly, maybe we can use those?

Reply:
yes, great point - I suggest doing that in the next step